File: python/src/vcard/validator.py
"""
vCard Validator for RFC 6350
"""

import re
from typing import List
from .dom import VCardObject


class ValidationResult:
    """Result of validation operation"""

    def __init__(self):
        self.errors: List[str] = []
        self.warnings: List[str] = []

    @property
    def is_valid(self) -> bool:
        """Check if validation passed (no errors)"""
        return len(self.errors) == 0

    def add_error(self, error: str):
        """Add an error message"""
        self.errors.append(error)

    def add_warning(self, warning: str):
        """Add a warning message"""
        self.warnings.append(warning)

    def get_summary(self) -> str:
        """Get a summary of validation results"""
        summary = f"Validation Result: {'VALID' if self.is_valid else 'INVALID'}\n"
        summary += f"Errors: {len(self.errors)}\n"
        summary += f"Warnings: {len(self.warnings)}\n"

        if self.errors:
            summary += "\nErrors:\n"
            for error in self.errors:
                summary += f"  - {error}\n"

        if self.warnings:
            summary += "\nWarnings:\n"
            for warning in self.warnings:
                summary += f"  - {warning}\n"

        return summary


class VCardValidator:
    """Validator for vCard components according to RFC 6350"""

    def validate(self, vcard: VCardObject) -> ValidationResult:
        """Validate a vCard object"""
        result = ValidationResult()
        self._validate_vcard(vcard, result)
        return result

    def _validate_vcard(self, vcard: VCardObject, result: ValidationResult):
        """Validate vCard properties"""
        # VCARD MUST have VERSION and FN
        self._validate_required_property(vcard, "VERSION", result)
        self._validate_required_property(vcard, "FN", result)

        # VERSION must be 4.0 (or 3.0, 2.1 for backward compatibility)
        version = vcard.get_property("VERSION")
        if version:
            valid_versions = ["4.0", "3.0", "2.1"]
            if version.value not in valid_versions:
                result.add_error(
                    f"VERSION must be one of: {', '.join(valid_versions)}, found: {version.value}"
                )
            elif version.value != "4.0":
                result.add_warning(
                    f"VERSION {version.value} is supported but deprecated. Consider upgrading to 4.0"
                )

        # FN (Formatted Name) must not be empty
        fn = vcard.get_property("FN")
        if fn and not fn.value.strip():
            result.add_error("FN (Formatted Name) cannot be empty")

        # Validate telephones
        for tel in vcard.telephones:
            self._validate_telephone(tel, result)

        # Validate emails
        for email in vcard.emails:
            self._validate_email(email, result)

        # Validate addresses
        for adr in vcard.addresses:
            self._validate_address(adr, result)

        # Validate URLs
        for url in vcard.urls:
            self._validate_url(url, result)

    def _validate_required_property(self, vcard: VCardObject, property_name: str, result: ValidationResult):
        """Validate that a required property is present"""
        prop = vcard.get_property(property_name)
        if not prop:
            result.add_error(f"Required property {property_name} is missing")

    def _validate_telephone(self, prop, result: ValidationResult):
        """Validate a telephone property"""
        if not prop.value.strip():
            result.add_error("TEL property cannot be empty")

        # Validate TYPE parameter if present
        types = prop.get_parameters("TYPE")
        valid_types = ["work", "home", "text", "voice", "fax", "cell", "video", "pager", "textphone"]
        for type_val in types:
            if type_val.lower() not in valid_types:
                result.add_warning(f"TEL TYPE parameter has non-standard value: {type_val}")

    def _validate_email(self, prop, result: ValidationResult):
        """Validate an email property"""
        if not prop.value.strip():
            result.add_error("EMAIL property cannot be empty")
            return

        # Basic email validation
        email_pattern = r'^[^@\s]+@[^@\s]+\.[^@\s]+$'
        if not re.match(email_pattern, prop.value):
            result.add_warning(f"EMAIL property may not be a valid email address: {prop.value}")

    def _validate_address(self, prop, result: ValidationResult):
        """Validate an address property"""
        # ADR property format: POBox;Extended;Street;Locality;Region;PostalCode;Country
        parts = prop.value.split(';')
        if len(parts) != 7:
            result.add_warning(
                f"ADR property should have exactly 7 components, found {len(parts)}: {prop.value}"
            )

    def _validate_url(self, prop, result: ValidationResult):
        """Validate a URL property"""
        if not prop.value.strip():
            result.add_error("URL property cannot be empty")
            return

        # Basic URL validation
        if not prop.value.startswith(('http://', 'https://', 'ftp://')):
            result.add_warning(f"URL property may not be a valid URL: {prop.value}")

File: typescript/src/validator.ts
/**
 * Validator for vCard components according to RFC 6350
 */

import { VCardComponent, VCardObject, VCardProperty } from './dom';

export class ValidationResult {
  errors: string[] = [];
  warnings: string[] = [];

  get isValid(): boolean {
    return this.errors.length === 0;
  }

  addError(error: string): void {
    this.errors.push(error);
  }

  addWarning(warning: string): void {
    this.warnings.push(warning);
  }

  getSummary(): string {
    let summary = `Validation Result: ${this.isValid ? 'VALID' : 'INVALID'}\n`;
    summary += `Errors: ${this.errors.length}\n`;
    summary += `Warnings: ${this.warnings.length}\n`;

    if (this.errors.length > 0) {
      summary += '\nErrors:\n';
      for (const error of this.errors) {
        summary += `  - ${error}\n`;
      }
    }

    if (this.warnings.length > 0) {
      summary += '\nWarnings:\n';
      for (const warning of this.warnings) {
        summary += `  - ${warning}\n`;
      }
    }

    return summary;
  }
}

export class VCardValidator {
  validate(vcard: VCardObject): ValidationResult {
    const result = new ValidationResult();
    this.validateVCard(vcard, result);
    return result;
  }

  private validateVCard(vcard: VCardObject, result: ValidationResult): void {
    // Required properties
    this.validateRequiredProperty(vcard, 'VERSION', result);
    this.validateRequiredProperty(vcard, 'FN', result);

    // VERSION must be 4.0 (or 3.0, 2.1 for backward compatibility)
    const versionProps = vcard.getProperties('VERSION');
    if (versionProps.length > 1) {
      result.addError('Duplicate VERSION property');
    }
    const version = vcard.getProperty('VERSION');
    if (version) {
      const validVersions = ['4.0', '3.0', '2.1'];
      if (!validVersions.includes(version.value)) {
        result.addError(`VERSION must be one of: ${validVersions.join(', ')}, found: ${version.value}`);
      } else if (version.value !== '4.0') {
        result.addWarning(`VERSION ${version.value} is supported but deprecated. Consider upgrading to 4.0`);
      }
    }

    // FN must not be empty; duplicate FN is an error
    const fnProps = vcard.getProperties('FN');
    if (fnProps.length > 1) {
      result.addError('Duplicate FN property');
    }
    const fn = vcard.getProperty('FN');
    if (fn && !fn.value.trim()) {
      result.addError('FN (Formatted Name) cannot be empty');
    }

    // Validate N format if present
    const name = vcard.getProperty('N');
    if (name) {
      this.validateStructuredName(name, result);
    }

    // Validate telephones
    for (const tel of vcard.telephones) {
      this.validateTelephone(tel, result);
    }
    for (const telProp of vcard.getProperties('TEL')) {
      this.validateTelProperty(telProp, result);
    }

    // Validate emails
    for (const email of vcard.emails) {
      this.validateEmail(email, result);
    }

    // Validate addresses
    for (const adr of vcard.addresses) {
      this.validateAddress(adr, result);
    }

    // Validate BDAY
    const bday = vcard.getProperty('BDAY');
    if (bday) {
      this.validateDateFormat(bday, result, 'BDAY');
    }

    // Validate ANNIVERSARY
    const anniversary = vcard.getProperty('ANNIVERSARY');
    if (anniversary) {
      this.validateDateFormat(anniversary, result, 'ANNIVERSARY');
    }

    // Validate GEO
    const geo = vcard.getProperty('GEO');
    if (geo) {
      this.validateGeoFormat(geo, result);
    }

    // Validate TZ
    const tz = vcard.getProperty('TZ');
    if (tz) {
      this.validateTimeZoneFormat(tz, result);
    }

    // Validate UID
    const uid = vcard.getProperty('UID');
    if (uid && !uid.value.trim()) {
      result.addError('UID cannot be empty if present');
    }

    // Validate MEMBER references (KIND:group requires UID for member resolution)
    const kind = vcard.getProperty('KIND');
    if (kind && kind.value.toLowerCase() === 'group') {
      const members = vcard.getProperties('MEMBER');
      const uid = vcard.getProperty('UID');
      if (members.length > 0 && !uid) {
        result.addError('KIND:group vCard with MEMBER properties must have a UID');
      }
      if (uid) {
        for (const member of members) {
          if (member.value === `urn:uuid:${uid.value}`) {
            result.addError('MEMBER property cannot reference the vCard itself (circular reference)');
          }
        }
      }
    }

    // Validate all properties for valid names and parameters
    for (const [name, props] of vcard.properties) {
      // Property names must be alphanumeric, hyphens, or dots (for groups)
      if (!/^[A-Za-z0-9._-]+$/.test(name)) {
        result.addError(`Invalid property name: ${name}`);
      }
      for (const prop of props) {
        // Check for invalid ENCODING parameter
        const encoding = prop.getParameter('ENCODING');
        if (encoding) {
          const validEncodings = ['B', 'BASE64', 'QUOTED-PRINTABLE', '8BIT'];
          if (validEncodings.includes(encoding.toUpperCase())) {
            result.addError(`ENCODING parameter is not valid in vCard 4.0: ${encoding}`);
          }
        }
      }
    }

    // Validate URLs
    for (const url of vcard.urls) {
      this.validateUrlFormat(url, result);
    }
  }

  private validateRequiredProperty(component: VCardComponent, propertyName: string, result: ValidationResult): void {
    if (!component.getProperty(propertyName)) {
      result.addError(`Required property ${propertyName} is missing`);
    }
  }

  private validateStructuredName(property: VCardProperty, result: ValidationResult): void {
    const parts = property.value.split(';');
    if (parts.length > 5) {
      result.addError(`N property has more than 5 components: ${property.value}`);
    }
    if (parts.every((p) => !p.trim())) {
      result.addError('N property must have at least one non-empty component');
    }
  }

  private validateTelephone(telephone: { value: string; types: number }, result: ValidationResult): void {
    if (!telephone.value.trim()) {
      result.addError('TEL property cannot be empty');
    }
  }

  private validateTelProperty(property: VCardProperty, result: ValidationResult): void {
    const validTelTypes = ['text', 'voice', 'fax', 'cell', 'video', 'pager', 'textphone', 'work', 'home'];
    for (const type of property.getParameters('TYPE')) {
      for (const t of type.split(',')) {
        if (!validTelTypes.includes(t.trim().toLowerCase())) {
          result.addError(`TEL TYPE parameter has invalid value: ${t}`);
        }
      }
    }
  }

  private validateEmail(email: { value: string; types: number }, result: ValidationResult): void {
    if (!email.value.trim()) {
      result.addError('EMAIL property cannot be empty');
      return;
    }
    const emailPattern = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
    if (!emailPattern.test(email.value)) {
      result.addWarning(`EMAIL property may not be a valid email address: ${email.value}`);
    }
  }

  private validateAddress(_address: { types: number }, _result: ValidationResult): void {
    // All address components are optional
  }

  private validateDateFormat(property: VCardProperty, result: ValidationResult, propertyName: string): void {
    const value = property.value;
    if (!value.trim()) {
      result.addError(`${propertyName} cannot be empty`);
      return;
    }

    const patterns = [
      /^\d{8}$/,                              // YYYYMMDD
      /^\d{4}-\d{2}-\d{2}$/,                  // YYYY-MM-DD
      /^--\d{4}$/,                            // --MMDD
      /^--\d{2}-\d{2}$/,                      // --MM-DD
      /^\d{4}-\d{2}$/,                        // YYYY-MM
      /^\d{4}$/,                              // YYYY
      /^\d{8}T\d{6}$/,                        // YYYYMMDDTHHMMSS
      /^\d{8}T\d{6}Z$/,                       // YYYYMMDDTHHMMSSZ
      /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}$/, // YYYY-MM-DDTHH:MM:SS
    ];

    if (!patterns.some((p) => p.test(value))) {
      result.addError(`${propertyName} has non-standard date format: ${value}`);
    }
  }

  private validateGeoFormat(property: VCardProperty, result: ValidationResult): void {
    const value = property.value;
    if (!value.trim()) {
      result.addError('GEO property cannot be empty');
      return;
    }
    const geoPattern = /^geo:[-+]?\d+\.?\d*,[-+]?\d+\.?\d*$/;
    if (!geoPattern.test(value)) {
      result.addWarning(`GEO property may not be in correct format (expected geo:lat,long): ${value}`);
    }
  }

  private validateTimeZoneFormat(property: VCardProperty, result: ValidationResult): void {
    const value = property.value;
    if (!value.trim()) {
      result.addError('TZ property cannot be empty');
      return;
    }
    const utcOffsetPattern = /^[+-]\d{2}:?\d{2}$/;
    if (!utcOffsetPattern.test(value) && !value.includes('/')) {
      result.addWarning(`TZ property may not be in correct format: ${value}`);
    }
  }

  private validateUrlFormat(property: VCardProperty, result: ValidationResult): void {
    const value = property.value;
    if (!value.trim()) {
      result.addError('URL property cannot be empty');
      return;
    }
    try {
      new URL(value);
    } catch {
      result.addError(`URL property is not a valid URL: ${value}`);
    }
  }
}
